File: app-updater/src/main/java/com/king/app/updater/http/IHttpManager.java
package com.king.app.updater.http;

import android.support.annotation.Nullable;

import java.io.File;
import java.io.Serializable;
import java.util.Map;

/**
 * @author Jenly <a href="mailto:jenly1314@gmail.com">Jenly</a>
 */
public interface IHttpManager {


    /**
     * 下载
     * @param url
     * @param path
     * @param filename
     * @param requestProperty
     * @param callback
     */
    void download(String url, String path, String filename, @Nullable Map<String,String> requestProperty, DownloadCallback callback);


    interface DownloadCallback extends Serializable{
        /**
         * 开始
         * @param url
         */
        void onStart(String url);

        /**
         * 加载进度…
         * @param progress
         * @param total
         */
        void onProgress(int progress,int total);

        /**
         * 完成
         * @param file
         */
        void onFinish(File file);

        /**
         * 错误
         * @param e
         */
        void onError(Exception e);


        /**
         * 取消
         */
        void onCancel();
    }
}

File: app-updater/src/main/java/com/king/app/updater/http/IHttpManager.kt
package com.king.app.updater.http

import java.io.File
import kotlinx.coroutines.flow.Flow

/**
 * Http下载管理器 默认提供 [HttpManager] 和 [OkHttpManager] 两种实现。
 *
 * @author <a href="mailto:jenly1314@gmail.com">Jenly</a>
 * <p>
 * <a href="https://github.com/jenly1314">Follow me</a>
 */
interface IHttpManager {

    /**
     * 下载
     *
     * @param url 下载地址
     * @param filepath 文件路径
     * @param headers 请求头
     */
    fun download(
        url: String,
        filepath: String,
        headers: Map<String, String>? = null,
    ): Flow<DownloadState>

    /**
     * 取消下载
     */
    fun cancel()

}

/**
 * 下载状态
 */
sealed class DownloadState {
    /**
     * 开始下载
     */
    data class Start(val url: String) : DownloadState()

    /**
     * 下载进度
     */
    data class Progress(
        val progress: Int,
        val total: Int,
    ) : DownloadState()

    /**
     * 下载成功
     */
    data class Success(val file: File) : DownloadState()

    /**
     * 下载出错
     */
    data class Error(val cause: Throwable) : DownloadState()

    /**
     * 下载取消
     */
    data object Cancel : DownloadState()
}
